src/linear_algebra/quadratic_form.lean

```diff
@@ -770,63 +772,54 @@ open finite_dimensional
   induction hd : finrank K V with d ih generalizing V,
-  { exact ⟨basis_of_finrank_zero hd, λ _ _ _, zero_left _, fin.elim0⟩ },
+  { exact ⟨basis_of_finrank_zero hd, λ _ _ _, zero_left _⟩ },
   haveI := finrank_pos_iff.1 (hd.symm ▸ nat.succ_pos d : 0 < finrank K V),
-  cases exists_bilin_form_self_neq_zero hB₁ hB₂ with x hx,
-  have hd' := hd,
-  rw [← submodule.finrank_add_eq_of_is_compl
-        (is_compl_span_singleton_orthogonal hx).symm,
+  -- either the bilinear form is trivial or we can pick a non-null `x`
+  obtain rfl | hB₁ := eq_or_ne B 0,
+  { let b := finite_dimensional.fin_basis K V,
+    rw hd at b,
+    refine ⟨b, λ i j hij, rfl⟩, },
+  obtain ⟨x, hx⟩ := exists_bilin_form_self_neq_zero hB₁ hB₂,
+  rw [← submodule.finrank_add_eq_of_is_compl (is_compl_span_singleton_orthogonal hx).symm,
       finrank_span_singleton (ne_zero_of_not_is_ortho_self x hx)] at hd,
-  rcases @ih (B.orthogonal $ K ∙ x) _ _ _
-    (B.restrict _) (B.restrict_orthogonal_span_singleton_nondegenerate hB₁ hB₂ hx)
-    (B.restrict_sym hB₂ _) (nat.succ.inj hd) with ⟨v', hv₁, hv₃⟩,
-
-  set v := (λ (i : fin _), if h : i = 0 then x else coe (v' (i.pred h))) with v_def,
-  have : ∀ i j (hij : i ≠ j), B.is_ortho (v i) (v j),
-  { intros i j hij,
-    simp only [v_def],
-    split_ifs with hi hj hj,
-    { have : i = j := hi.trans hj.symm, contradiction },
-    { exact (v' (j.pred hj)).2 _ (submodule.mem_span_singleton_self x) },
+  let B' := B.restrict (B.orthogonal $ K ∙ x),
+  obtain ⟨v', hv₁⟩ := ih (B.restrict_sym hB₂ _ : sym_bilin_form.is_sym B') (nat.succ.inj hd),
+  -- concatenate `x` with the basis obtained by induction
+  let b := basis.mk_fin_cons x v'
+    (begin
+      rintros c y hy hc,
+      rw add_eq_zero_iff_neg_eq at hc,
+      rw [← hc, submodule.neg_mem_iff] at hy,
+      have := (is_compl_span_singleton_orthogonal hx).disjoint,
+      rw submodule.disjoint_def at this,
+      have := this (c • x) (submodule.smul_mem _ _ $ submodule.mem_span_singleton_self _) hy,
+      exact (smul_eq_zero.1 this).resolve_right (λ h, hx $ h.symm ▸ zero_left _),
+    end)
+    (begin
+      intro y,
+      refine ⟨-B x y/B x x, λ z hz, _⟩,
+      obtain ⟨c, rfl⟩ := submodule.mem_span_singleton.1 hz,
+      rw [is_ortho, smul_left, add_right, smul_right, div_mul_cancel _ hx, add_neg_self, mul_zero],
+    end),
+  refine ⟨b, _⟩,
+  { rw basis.coe_mk_fin_cons,
+    intros j i,
+    refine fin.cases _ (λ i, _) i; refine fin.cases _ (λ j, _) j; intro hij;
+      simp only [fin.cons_zero, fin.cons_succ, function.comp_apply],
+    { exact (hij rfl).elim },
+    { exact (v' j).prop _ (submodule.mem_span_singleton_self x) },
     { rw [is_ortho, hB₂],
-      exact (v' (i.pred hi)).2 _ (submodule.mem_span_singleton_self x) },
-    { exact hv₁ (j.pred hj) (i.pred hi) (by simpa using hij.symm) } },
-
-  refine ⟨@basis_of_linear_independent_of_card_eq_finrank _ _ _ _ _ _ _ _
-      v
-      (@linear_independent_of_is_Ortho _ _ _ _ _ _ B v (λ i j hij, this j i hij.symm) _)
```

Comment:
This is the only place that `linear_independent_of_is_Ortho` was used. Perhaps this suggests it too can be generalized.

Reply:
Right now it needs a hypothesis that the basis has no "null" vectors, but evidently in this situation we were still able to show linear independence another way